permissions/safe-router.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * SAFE ROUTER
+ * This router ensures that an endpoint is only run if any middleware has authorized it previously.
+ *
+ * How does it work?
+ * 1) It first runs a middleware that mark the request as not authorized yet.
+ * 2) Just before the endpoint is run, another middleware checks if the request has been authorized.
+ *
+ * How to use it?
+ * It's similar to Express Router (it actually uses it internally):
+ *
+ * const safeRouter = SafeRouter();
+ * safeRouter.get('/:id', customPermissionMiddlewares, () => {});
+ * safeRouter.post('/search', customPermissionMiddlewares, () => {});
+ * // Integrate with Express using SafeRouter's field 'expressRouter'
+ * router.use('/resource', safeRouter.expressRouter);
+ *
+ * IMPORTANT NOTE:
+ * Any authorization middleware should set 'req.authorized' to true and call next()
+ * to mark the request as authorized.
+ * DO NOT set 'req.authorized' to false NEVER. In case a middleware cannot authorize the request, just
```

Comment:
This would be easy to work around it. For example, instead of changing the value of `req.authorized` we could just have a method `req.authorize()` and call it in the middlewares. We could choose to have or not a `req.notAuhtorized()` method. This way, the user could never misuse SafeRouter.

Reply:
I really like the idea of having this `req.authorize` method.